[SOLID] Change if && to early return

dobryy opened the pull request.

dobryy commented:
Just checking out very basic logic...

TomasVotruba commented on rules/nette/src/TemplatePropertyAssignCollector.php:
```diff
@@ -100,11 +100,13 @@ private function collectVariableFromAssign(Assign $assign): void
             $this->nodesToRemove[] = $assign;
         }
-
-        // $x = $this->template
```
This comment should be kept

TomasVotruba commented on rules/solid/src/Rector/If_/ChangeAndIfToEarlyReturnRector.php:
```diff
@@ -0,0 +1,227 @@
+        $secondIf = new If_($invertedRightCondition);
+        $secondIf->stmts = [new Return_()];
+
+        $this->addNodeAfterNode($firstIf, $node);
+        $this->addNodeAfterNode($secondIf, $node);
+        $this->addNodeAfterNode($ifReturn, $node);
```
These is also `addNodesAfterNode()` method

TomasVotruba commented on rules/solid/tests/Rector/If_/ChangeAndIfToEarlyReturnRector/Fixture/equals.php.inc:
```diff
@@ -0,0 +1,37 @@
+{
+    public function canDrive(Car $car)
+    {
+        if ($car->hasWheels === true && $car->hasFuel === true) {
```
Could you also add test for 3 items?

```php
$value === true && $value === true && $value === true 
```

dobryy replied:
At the moment, the case with 3 items is skipped, there is a fixture for this. More than two items will make rule a bit more complicated I think. I wanted to keep the PR as small as possible.

TomasVotruba replied:
I see

dobryy commented on rules/solid/tests/Rector/If_/ChangeAndIfToEarlyReturnRector/Fixture/skip_more_than_two_conditions.php.inc:
```diff
@@ -0,0 +1,17 @@
+{
+    public function canDrive(Car $car)
+    {
+        if ($car->hasWheels && $car->hasFuel && $car->hasEngine) {
```
@TomasVotruba This situation can be handled in the next PR.

dobryy commented on rules/solid/tests/Rector/If_/ChangeAndIfToEarlyReturnRector/Fixture/skip_non_first_level_if.php.inc:
```diff
@@ -0,0 +1,21 @@
+{
+    public function canDrive(Car $car)
+    {
+        if ($car->wheelsCount() > 2) {
+            if ($car->hasWheels && $car->hasFuel) {
+                $this->canDrive = true;
+            }
+
+            return;
+        }
+
+        $this->canDrive = true;
```
@TomasVotruba This situation can be handled in the next PR.

dobryy commented on rules/solid/tests/Rector/If_/ChangeAndIfToEarlyReturnRector/Fixture/skip_if_in_foreach.php.inc:
```diff
@@ -0,0 +1,24 @@
+{
+    private function refactorStmts(array $nodes): void
+    {
+        foreach ($nodes as $node) {
+            if ($node instanceof Name && $node instanceof Identifier) {
+                $this->processNameOrIdentifier($node);
+            }
+
+            return;
+        }
+
```
@TomasVotruba This situation looks crazy but also can be handled in the next PR.

TomasVotruba reviewed (changes requested):
I've added few comments

dobryy commented:
> I've added few comments

I've implemented changes as a new commit https://github.com/rectorphp/rector/pull/4344/commits/38047f43579ff5916c14f97fbd122002bb0464b9

dobryy commented:
@TomasVotruba Pipeline is finally green.

TomasVotruba commented:
Thank you

ERuban commented:
@dobryy @TomasVotruba hey, I've just a one question - why it is a SOLID rule?

TomasVotruba commented:
From all the sets there are now, SOLID and early return are closest.
What set would you suggest instead?

ERuban commented:
@TomasVotruba So, I would suggest `CodingStyle` set for this rule - seems it will be better here, just because it is really more about accepted conventions in teams/companies, imho.
Also, why do you think it is closest to SOLID rules?

ERuban commented:
Hm, it will be really nice to move it to a standalone set. 
Or, maybe, to CodeQuality set :)

TomasVotruba commented:
Then, next month we a question, why it's part of CodeQuality set :D 

Btw, is current situation causing any undesired behavior of Rector in your code or is it rather academic curiosity?

TomasVotruba commented:
There is an issue here: https://github.com/rectorphp/rector/issues/4464

ERuban commented:
So, ok, standalone set looks ok, imo. 
Yeap, it is just academic curiosity (we haven't such rule in our team). 

Btw, I've some issues after updating Rector from the older version, but will ask you separately about it. ;)
